src/time_bucket_ng.c

```diff
@@ -0,0 +1,154 @@
+/*
+ * This file and its contents are licensed under the Apache License 2.0.
+ * Please see the included NOTICE for copyright information and
+ * LICENSE-APACHE for a copy of the license.
+ */
+#include <postgres.h>
+#include <utils/date.h>
+#include <utils/datetime.h>
+#include <utils/fmgrprotos.h>
+
+#include "time_bucket_ng.h"
+
+TS_FUNCTION_INFO_V1(ts_time_bucket_ng);
+TS_FUNCTION_INFO_V1(ts_time_bucket_ng_timestamp);
+TS_FUNCTION_INFO_V1(ts_time_bucket_ng_timestamptz);
+
+TSDLLEXPORT Datum
+ts_time_bucket_ng_timestamp(PG_FUNCTION_ARGS)
+{
+	DateADT result;
+	Datum interval = PG_GETARG_DATUM(0);
+	DateADT ts_date = DatumGetDateADT(DirectFunctionCall1(timestamp_date, PG_GETARG_DATUM(1)));
+
+	if (PG_NARGS() > 2)
+	{
+		DateADT origin = DatumGetDateADT(DirectFunctionCall1(timestamp_date, PG_GETARG_DATUM(2)));
+		result = DatumGetDateADT(DirectFunctionCall3(ts_time_bucket_ng,
+													 interval,
+													 DateADTGetDatum(ts_date),
+													 DateADTGetDatum(origin)));
+	}
+	else
+	{
+		result = DatumGetDateADT(
+			DirectFunctionCall2(ts_time_bucket_ng, interval, DateADTGetDatum(ts_date)));
+	}
+
+	return DirectFunctionCall1(date_timestamp, DateADTGetDatum(result));
+}
+
+TSDLLEXPORT Datum
+ts_time_bucket_ng_timestamptz(PG_FUNCTION_ARGS)
+{
+	DateADT result;
+	Datum interval = PG_GETARG_DATUM(0);
+	DateADT ts_date = DatumGetDateADT(DirectFunctionCall1(timestamptz_date, PG_GETARG_DATUM(1)));
+
+	if (PG_NARGS() > 2)
+	{
+		DateADT origin = DatumGetDateADT(DirectFunctionCall1(timestamptz_date, PG_GETARG_DATUM(2)));
+		result = DatumGetDateADT(DirectFunctionCall3(ts_time_bucket_ng,
```

Comment:
Should probably have a test for this code path.

Reply:
Good point. Fixed!